src/PrestaShopBundle/EventListener/ModuleActivatedListener.php

```diff
@@ -0,0 +1,198 @@
+
+    /**
+     * @param FilterControllerEvent $event
+     *
+     * @throws AnnotationException
+     * @throws \ReflectionException
+     */
+    public function onKernelController(FilterControllerEvent $event)
+    {
+        if (!$event->isMasterRequest()) {
+            return;
+        }
+
+        $controller = $event->getController();
+
+        if (!is_array($controller)) {
+            return;
+        }
+
+        list($controllerObject, $methodName) = $controller;
+        $moduleActivated = $this->getAnnotation($controllerObject, $methodName);
+
+        if (null === $moduleActivated) {
+            return;
+        }
+
+        /** @var \Module $module */
+        $module = \Module::getInstanceByName($moduleActivated->getModuleName());
+        if (null === $module) {
+            return;
+        }
+
+        if (!$module->active) {
+            $this->showNotificationMessage($moduleActivated);
+            $url = $this->router->generate($moduleActivated->getRedirectRoute());
+
+            $event->setController(function () use ($url) {
+                return new RedirectResponse($url);
+            });
+        }
+    }
+
+    /**
+     * Send an error message when redirected, will only work on migrated pages.
+     *
+     * @param ModuleActivated $moduleActivated
+     */
+    private function showNotificationMessage(ModuleActivated $moduleActivated)
+    {
+        $this->session->getFlashBag()->add(
+            'error',
+            $this->translator->trans(
+                $moduleActivated->getMessage(),
+                [$moduleActivated->getModuleName()],
+                $moduleActivated->getDomain()
+            )
+        );
+    }
+
+    /**
+     * @param object $controllerObject
+     * @param string $methodName
+     *
+     * @return null|ModuleActivated
+     * @throws AnnotationException
+     * @throws \ReflectionException
+     */
+    private function getAnnotation($controllerObject, $methodName)
+    {
+        $tokenAnnotation = ModuleActivated::class;
+
+        $controllerClass = ClassUtils::getClass($controllerObject);
+        $classAnnotation = $this->annotationReader->getClassAnnotation(
+            new ReflectionClass($controllerClass),
+            $tokenAnnotation
+        );
+
+        if (null !== $classAnnotation && $classAnnotation instanceof ModuleActivated) {
+            $this->validateAnnotation($classAnnotation, $controllerClass);
+
+            return $classAnnotation;
+        }
+
+        $controllerReflectionObject = new ReflectionObject($controllerObject);
+        $reflectionMethod = $controllerReflectionObject->getMethod($methodName);
+
+        $annotation = $this->annotationReader->getMethodAnnotation($reflectionMethod, $tokenAnnotation);
+
+        if (null !== $annotation && $annotation instanceof ModuleActivated) {
+            $this->validateAnnotation($annotation, $controllerClass . '::' . $methodName);
+
+            return $annotation;
+        }
+
+        return null;
```

Comment:
you don't need it :)

Reply:
in the (weird) case `$annotation` is not null but is not a `ModuleActivated` instance

Reply:
:'( FML

Reply:
true, it's a habit I have for methods which are supposed to return a value to always return something, even when it returns null
as opposition to methods with no return, like in C where function return `void`
It has no real use except for carity

Reply:
fair enough :+1: